### PR TITLE
Build and attach Javadoc artifact on release

### PR DESCRIPTION
**Pull request by joschi**
This PR creates and attaches the Javadoc artifact for Elasticsearch (core) during a release and will upload that artifact to the Maven repository, along with the original binary artifact and the source artifact.

This change would enable users and developers use a service like http://www.javadoc.io/ to read the Elasticsearch Javadoc instead of having to build it themselves or read the complete source code.

If this PR is being merged (or at least a candidate for merging), I would port it to the Gradle build in the `master` branch as well.

Refs #1203

### CODE REVIEW

File: core/pom.xml [hunk 1]
```diff
@@ -370,5 +370,28 @@
...
+                        <groupId>org.apache.maven.plugins</groupId>
+                        <artifactId>maven-javadoc-plugin</artifactId>
+                        <executions>
+                            <execution>
+                                <id>attach-javadocs</id>
+                                <goals>
+                                    <goal>jar</goal>
+                                </goals>
+                                <configuration>
+                                    <additionalparam>-Xdoclint:none</additionalparam>
```
Comment (rjernst) [inline core/pom.xml]: We should not be disabling doclint. We are very careful to fully enable it, except for "missing" docs, in our javac settings.

**Comment by rjernst** [comment]
I left one comment. Also, this has already been added in master for gradle in #14766

File: core/pom.xml (hunk 1 above)
Comment (joschi) [inline core/pom.xml]: I would agree but there are many linter errors in the current sources which would prevent build the artifact (at least on Java 8).

Would it be okay to just disable the Javadoc doclint on Java 8, e. g. as described in http://blog.joda.org/2014/02/turning-off-doclint-in-jdk-8-javadoc.html?

Comment (rjernst) [inline core/pom.xml]: I think we should backport the fixes that I made to get doclint fully passing on master. I found that, although we run doclint in javac, it appears it misses some issues which javadoc catches. I can backport the fixes from the PR I referenced tomorrow, or you can take a stab at backporting or merging them into this PR.

Comment (joschi) [inline core/pom.xml]: I took a stab at fixing the doclint errors but quite frankly, it's just too many right now to fix in the scope of this PR. Some of them can be fixed mechanically (e. g. with `sed`) but many more require manual intervention.

Take a look at https://github.com/joschi/elasticsearch/tree/publish-javadoc-artifact-including-javadoc-fixes for some of the fixes.

**Comment by joschi** [comment]
Rebased on current `2.x` branch.

File: core/pom.xml (hunk 1 above)
Comment (rjernst) [inline core/pom.xml]: The errors have already been fixed on master (see #13702). There was a big cleanup, it was not just the few fixes in #14766.  You can try cherry picking them; there were 10 or so files that needed manual merging when I glanced at it. But I still would not want to push this change without doclint: we need to ensure we are building good javadocs.

**Comment by joschi** [comment]
Rebased on current 2.x branch and pulled in Javadoc fixes from #13702 and #14766.

**Comment by s1monw** [comment]
@joschi thanks a lot for the effort, would it be possible to only commit the pom.xml file change and make it a bit more lenient or at least also do these changes on master? @rjernst WDYT?

**Comment by joschi** [comment]
@s1monw Thanks for coming back to this PR.

The changes of the POM on themselves aren't too useful because it'll make the build fail due to the Javadoc lint errors that Java 8 will emit (not a problem on Java 7, though), see the previous comments by @rjernst (https://github.com/elastic/elasticsearch/pull/14935#discussion_r45709095).

The second commit pulls in the Javadoc fixes from #13702 and #14766 so that the build won't fail (can be checked by simply running `mvn javadoc:javadoc`). Those have already been applied to `master` so that this PR is purely for the 2.x line of Elasticsearch.

**Comment by s1monw** [comment]
@joschi I am really wondering if we should go through the trouble and fix it in 2.x. It's happening in master and maybe we should just go with what we have to reduce the noise?

**Comment by joschi** [comment]
@s1monw In the end it's your call. In my opinion having publicly available Javadoc for Elasticsearch 2.x is worth the relatively minor changes (and it's only Javadoc…) in this PR.

**Comment by s1monw** [comment]
fair enough @rjernst I am ok with it, WDYT?

**Comment by rjernst** [comment]
@joschi I'm happy to merge this, but I see some problems in the diff. It looks like some changes that were meant for master only were inadvertently backported. For example, look at TransportShardMultiPercolateAction.java. Your patch has this:

```
-                if (in.getVersion().onOrAfter(Version.V_2_3_0)) {
-                    shardRequest.startTime(in.readLong());
-                }
```

Does running `mvn verify` work for you? The patch is so large I'm concerned other unintentional changes could be backported that could have similar bad results. @s1monw Because of this, I am still not sure if it is worth the risk to backport.

**Comment by joschi** [comment]
@rjernst I can't find any change of that file (`core/src/main/java/org/elasticsearch/action/percolate/TransportShardMultiPercolateAction.java`) in my branch compared to the `2.x` branch:
```
$ git rev-parse --abbrev-ref HEAD
publish-javadoc-artifact
$ git diff 2.x..publish-javadoc-artifact ./core/src/main/java/org/elasticsearch/action/percolate/TransportShardMultiPercolateAction.java
$
```

Maybe you're using an outdated state of this branch?

**Comment by javanna** [comment]
I remember reviewing such a change a few days ago, I think it was #15938, so the change has not been introduced with this PR.

**Comment by rjernst** [comment]
Doh! Sorry for the confusion! The script I use to checkout a PR messed this up. I've reviewed now and I will merge shortly.

**Pull request merged**